academy/ibc/ibc-tao-dev.md

```diff
@@ -0,0 +1,783 @@
+
+```go
+// ConnectionOpenInit defines a rpc handler method for MsgConnectionOpenInit.
+func (k Keeper) ConnectionOpenInit(goCtx context.Context, msg *connectiontypes.MsgConnectionOpenInit) (*connectiontypes.MsgConnectionOpenInitResponse, error) {
+  ctx := sdk.UnwrapSDKContext(goCtx)
+
+  if _, err := k.ConnectionKeeper.ConnOpenInit(ctx, msg.ClientId, msg.Counterparty, msg.Version, msg.DelayPeriod); err != nil {
+    return nil, sdkerrors.Wrap(err, "connection handshake open init failed")
+  }
+
+  return &connectiontypes.MsgConnectionOpenInitResponse{}, nil
+}
+```
+
+Now take a look at the `ConnOpenTry`:
+
+```go
+// ConnOpenTry relays notice of a connection attempt on chain A to chain B (this
+// code is executed on chain B).
+//
+// NOTE:
+//  - Here chain A acts as the counterparty
+//  - Identifiers are checked on msg validation
+func (k Keeper) ConnOpenTry(
+  ctx sdk.Context,
+  previousConnectionID string, // previousIdentifier
+  counterparty types.Counterparty, // counterpartyConnectionIdentifier, counterpartyPrefix and counterpartyClientIdentifier
+  delayPeriod uint64,
+  clientID string, // clientID of chainA
+  clientState exported.ClientState, // clientState that chainA has for chainB
+  counterpartyVersions []exported.Version, // supported versions of chain A
+  proofInit []byte, // proof that chainA stored connectionEnd in state (on ConnOpenInit)
+  proofClient []byte, // proof that chainA stored a light client of chainB
+  proofConsensus []byte, // proof that chainA stored chainB's consensus state at consensus height
+  proofHeight exported.Height, // height at which relayer constructs proof of A storing connectionEnd in state
+  consensusHeight exported.Height, // latest height of chain B which chain A has stored in its chain B client
+) ...
+```
+
+and `ConnOpenAck`:
+
+```go
+func (k Keeper) ConnOpenAck(
+  ctx sdk.Context,
+  connectionID string,
+  clientState exported.ClientState, // client state for chainA on chainB
+  version *types.Version, // version that ChainB chose in ConnOpenTry
+  counterpartyConnectionID string,
+  proofTry []byte, // proof that connectionEnd was added to ChainB state in ConnOpenTry
+  proofClient []byte, // proof of client state on chainB for chainA
+  proofConsensus []byte, // proof that chainB has stored ConsensusState of chainA on its client
+  proofHeight exported.Height, // height that relayer constructed proofTry
+  consensusHeight exported.Height, // latest height of chainA that chainB has stored on its chainA client
+) ...
+```
+
+both will do the same checks:
+
+```go
+
+  // Check that ChainA committed expectedConnectionEnd to its state
+  if err := k.VerifyConnectionState(
+    ctx, connection, proofHeight, proofInit, counterparty.ConnectionId,
+    expectedConnection,
+  ); err != nil {
+    return "", err
```

Comment:
explain difference between VerifyConnection/Client/ClientConsensusState.

from the internal recording:

these states come from stored states in the light clients which need to be updated by the relayers.

 VerifyConnection verifies that the photo they have looks like us (explain this in state terms)
 VerifyClient/Consensus verifies that the photo we have looks like them (explain this in state terms)